src/views/Roles/Roles.js

```diff
@@ -0,0 +1,408 @@
+  const assigning = useSelector((state) => state.statuses.assigning);
+  const updating = useSelector((state) => state.statuses.updating);
+  const deleting = useSelector((state) => state.statuses.deleting);
+  const users = useSelector((state) => state.users.users);
+  const roles = useSelector((state) => state.roles.roles);
+  const alert = useSelector((state) => state.alerts);
+
+  const toggle = () => setModal(!modal);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    document.title = "User Role Settings | Barefoot Nomad";
+    dispatch(getUsers());
+    dispatch(getRoles());
+  }, []);
+
+  useEffect(() => {
+    setEmail(users[0]?.email);
+  }, [users]);
+
+  useEffect(() => {
+    setUserRole(roles[0]?.name);
+  }, [roles]);
+
+  useEffect(() => {
+    if (!updating && modal) {
+      toggle();
+    }
+  }, [updating]);
+
+  useEffect(() => {
+    if (!deleting && modal) {
+      toggle();
+    }
+  }, [deleting]);
+
+  useEffect(() => {
+    if (!creating) {
+      setName("");
+      setDescription("");
+    }
+  }, [creating]);
+
+  useEffect(() => {
+    if (description.length >= 10 && name.length >= 3) {
+      setDisabled(false);
+    } else {
+      setDisabled(true);
+    }
+  }, [name, description]);
+
+  useEffect(() => {
+    if (name.length >= 3) {
+      setValidName("is-valid");
+    } else {
+      setValidName("is-invalid");
+    }
+  }, [name]);
+
+  useEffect(() => {
+    if (description.length >= 10) {
+      setValidDescription("is-valid");
+    } else {
+      setValidDescription("is-invalid");
+    }
+  }, [description]);
+
+  useEffect(() => {
+    setVisible(true);
+  }, [alert]);
+
+  const handleCreate = (e) => {
+    e.preventDefault();
+
+    dispatch(createRole({ name, description }));
+  };
+
+  const handleAssign = (e) => {
+    e.preventDefault();
+    dispatch(assignRole({ email, userRole }));
+  };
+
+  const findRoleName = (id) => {
+    const userRole = roles.filter((role) => role.id === id)[0];
+
+    return userRole ? userRole.name : "";
+  };
+
+  const openEditModal = (role) => {
+    setRoleInModal(role);
+    setModalName(role.name);
+    setModalDescription(role.description);
+    setWhichModal("edit");
+    toggle();
+  };
+
+  const openDeleteModal = (role) => {
+    setRoleInModal(role);
+    setWhichModal("delete");
+    toggle();
+  };
+
+  const isValid = ({ target }) => {
+    if (target.value.length < 3) {
+      target.classList.add("is-invalid");
+      target.classList.remove("is-valid");
+    } else {
+      target.classList.remove("is-invalid");
+      target.classList.add("is-valid");
+    }
+  };
+
+  const handleEdit = () => {
+    if (modalName.length < 3 || modalDescription.length < 3) {
+      return;
+    } else {
+      dispatch(
+        updateRole(roleInModal.id, {
+          name: modalName,
+          description: modalDescription,
+        })
+      );
+    }
+  };
+
+  const handleDelete = () => {
+    dispatch(deleteRole(roleInModal));
+  };
+
```

Comment:
this file so large please break it down, I could suggest you separate the logic from the views/components

Reply:
Fixed 👍